File: src/main/kotlin/io/energyconsumptionoptimizer/mapservice/application/HouseMapServiceImpl.kt
package io.energyconsumptionoptimizer.mapservice.application

import io.energyconsumptionoptimizer.mapservice.domain.HouseMap
import io.energyconsumptionoptimizer.mapservice.domain.errors.FlorPlanFormatNotFoundException
import io.energyconsumptionoptimizer.mapservice.domain.ports.FloorPlanService
import io.energyconsumptionoptimizer.mapservice.domain.ports.HouseMapService
import io.energyconsumptionoptimizer.mapservice.domain.ports.SmartFurnitureHookupService
import io.energyconsumptionoptimizer.mapservice.domain.ports.ZoneService

class HouseMapServiceImpl(
    private val floorPlanService: FloorPlanService,
    private val zoneService: ZoneService,
    private val smartFurnitureHookupService: SmartFurnitureHookupService,
) : HouseMapService {
    override suspend fun getHouseMap(): HouseMap {
        val florPlan = floorPlanService.getFloorPlan() ?: throw FlorPlanFormatNotFoundException()
        val zones = zoneService.getZones()
        val smartFurnitureHookups = smartFurnitureHookupService.getSmartFurnitureHookups()

        return HouseMap(florPlan, zones, smartFurnitureHookups)
    }
}

File: src/application/HouseMapServiceImpl.ts
import type { HouseMapService } from "@application/inbound/HouseMapService";
import type { FloorPlanService } from "@application/inbound/FloorPlanService";
import type { ZoneService } from "@application/inbound/ZoneService";
import type { SmartFurnitureHookupService } from "@application/inbound/SmartFurnitureHookupService";
import { HouseMap } from "@domain/entities/Map";
import { FloorPlanNotFoundError } from "@domain/errors";

export class HouseMapServiceImpl implements HouseMapService {
  readonly #floorPlanService: FloorPlanService;
  readonly #zoneService: ZoneService;
  readonly #smartFurnitureHookupService: SmartFurnitureHookupService;

  constructor(
    floorPlanService: FloorPlanService,
    zoneService: ZoneService,
    smartFurnitureHookupService: SmartFurnitureHookupService,
  ) {
    this.#floorPlanService = floorPlanService;
    this.#zoneService = zoneService;
    this.#smartFurnitureHookupService = smartFurnitureHookupService;
  }

  async getHouseMap(): Promise<HouseMap | Error> {
    const floorPlan = await this.#floorPlanService.getFloorPlan();
    if (!floorPlan) return new FloorPlanNotFoundError();

    const [zones, hookups] = await Promise.all([
      this.#zoneService.getZones(),
      this.#smartFurnitureHookupService.getSmartFurnitureHookups(),
    ]);

    return HouseMap.rehydrate(floorPlan, zones, hookups);
  }
}
